variant/effect.go

```diff
@@ -0,0 +1,276 @@
+
+// Effector is satisfied by types that implement the Effect method, which returns
+// a CodingChange describing how changes prescribed by the receiver type alter the
+// resulting amino acid sequence corresponding to the input []dna.Base. The input
+// sequence should be the CDS of a gene such that input[0:3] is the start codon.
+// The last 3 dna.Base of the input is not required to be a stop codon. Trailing
+// sequence after the stop codon (i.e. the 3' UTR) may be optionally added to the
+// input sequence, however the deletion position must be within the coding sequence.
+// This trailing sequence is used when a variant causes a frameshift and a new stop
+// codon in not found prior to the original stop codon. In this case the translation
+// will continue into the 3' UTR attempting to find a new stop codon.
+//
+// The offset integer input for the Effect method is added to the position in
+// the receiver type and may be positive or negative. Often this offset will
+// be the negative genomic position of the beginning of the start codon.
+//
+// Effector is satisfied by Substitution, Insertion, Deletion, and Delins.
+// The effects of Structural variation are often too complex to be described by
+// CodingChange and therefore should be handled separately.
+type Effector interface {
+	Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error)
+}
+
+// Effect for Substitution determines a coding change resulting from a substitution.
+func (s Substitution) Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error) {
+	var answer CodingChange
+	offsetPos := s.Pos + offsetStart
+	answer.CodingPos = offsetPos
+	answer.ProteinPos = offsetPos / 3
+
+	if offsetPos < 0 {
+		return answer, ErrNegPos
+	}
+
+	if codingSeq[offsetPos] != s.Ref {
+		return answer, ErrRefMatch
+	}
+
+	frame := offsetPos % 3
+	codonStart := offsetPos - frame
+
+	codon := dna.Codon{codingSeq[codonStart], codingSeq[codonStart+1], codingSeq[codonStart+2]}
+	refAa := dna.GeneticCode[codon]
+
+	codon[frame] = s.Alt // mutate codon
+	altAa := dna.GeneticCode[codon]
+
+	if refAa != altAa {
+		answer.RemovedAa = []dna.AminoAcid{refAa}
+		answer.AddedAa = []dna.AminoAcid{altAa}
+	}
+
+	switch {
+	case altAa == refAa:
+		answer.Type = Silent
+
+	case altAa == dna.Stop:
+		answer.Type = Nonsense
+
+	default:
+		answer.Type = Missense
+	}
+
+	return answer, nil
+}
+
+// Effect for Insertion determines a coding change resulting from an insertion.
+func (i Insertion) Effect(codingSeq []dna.Base, offsetStart int, offsetEnd int) (CodingChange, error) {
+	var answer CodingChange
+	offsetPos := i.Pos + offsetStart
+	answer.CodingPos = offsetPos
+	answer.ProteinPos = offsetPos / 3
+	var protOffset int
+
+	if offsetPos < 0 {
+		return answer, ErrNegPos
+	}
+
+	if offsetPos > len(codingSeq) {
+		return answer, ErrInvalidPosition
+	}
+
+	frame := offsetPos % 3
+	codonStart := offsetPos - frame
+
+	switch {
+	case len(i.Seq)%3 != 0: // frameshift
```

Comment:
I realize I am hitting corner cases, but do we know this is a frameshift at this point?  I am imagining the case where I insert a series of bases that is equal to the bases that come after the insertion (tandem duplication; protein seq remains identical).  It seems that to have a frameshift, we have to know that the insertion is not bringing its own stop codon.  Maybe we know that already, or there is a catch for it later on?

Reply:
I think I see what you mean. So if we have a 301bp insertion that duplicates the last 301bp of the gene then, while technically it creates a frameshift, the protein sequence is actually unchanged so we would want to report it differently. I have added a new check for this after the switch case to determine if the effect is silent.